src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,122 +287,202 @@ class HandshakeALotClosure : public HandshakeClosure {
+  log_debug(vmthread)("Evaluating %s %s VM operation: %s",
+                       prev_vm_operation != NULL ? "nested" : "",
+                      _cur_vm_operation->evaluate_at_safepoint() ? "safepoint" : "non-safepoint",
+                      _cur_vm_operation->name());
 
-      // If we are at a safepoint we will evaluate all the operations that
-      // follow that also require a safepoint
-      if (_cur_vm_operation->evaluate_at_safepoint()) {
-        log_debug(vmthread)("Evaluating safepoint VM operation: %s", _cur_vm_operation->name());
+  bool end_safepoint = false;
+  if (_cur_vm_operation->evaluate_at_safepoint() &&
+      !SafepointSynchronize::is_at_safepoint()) {
+    SafepointSynchronize::begin();
+    if (_timeout_task != NULL) {
+      _timeout_task->arm();
+    }
+    end_safepoint = true;
+  }
 
-        SafepointSynchronize::begin();
+  evaluate_operation(_cur_vm_operation);
 
-        if (_timeout_task != NULL) {
-          _timeout_task->arm();
-        }
+  if (end_safepoint) {
+    if (_timeout_task != NULL) {
+      _timeout_task->disarm();
+    }
+    SafepointSynchronize::end();
+  }
 
-        evaluate_operation(_cur_vm_operation);
-        _cur_vm_operation = NULL;
+  _cur_vm_operation = prev_vm_operation;
+}
 
-        if (_timeout_task != NULL) {
-          _timeout_task->disarm();
-        }
+void VMThread::wait_for_operation() {
+  MonitorLocker ml_op_lock(VMOperation_lock, Mutex::_no_safepoint_check_flag);
 
-        // Complete safepoint synchronization
-        SafepointSynchronize::end();
+  // Clear previous operation.
+  // On first call this clears a dummy place-holder.
+  _next_vm_operation = NULL;
+  // Notify operation done and notify a next operation can be installed.
+  ml_op_lock.notify_all();
 
-      } else {  // not a safepoint operation
-        log_debug(vmthread)("Evaluating non-safepoint VM operation: %s", _cur_vm_operation->name());
-        evaluate_operation(_cur_vm_operation);
-        _cur_vm_operation = NULL;
+  while (!should_terminate()) {
+    self_destruct_if_needed();
+    if (_next_vm_operation != NULL) {
+      return;
+    }
+    if (handshake_alot()) {
+      {
+        MutexUnlocker mul(VMOperation_lock);
+        HandshakeALotClosure hal_cl;
+        Handshake::execute(&hal_cl);
+      }
+      // When we unlocked above someone might have setup a new op.
+      if (_next_vm_operation != NULL) {
+        return;
       }
     }
+    assert(_next_vm_operation == NULL, "Must be");
+    assert(_cur_vm_operation  == NULL, "Must be");
 
-    //
-    //  Notify (potential) waiting Java thread(s)
-    { MonitorLocker mu(VMOperationRequest_lock, Mutex::_no_safepoint_check_flag);
-      mu.notify_all();
+    setup_periodic_safepoint_if_needed();
+    if (_next_vm_operation != NULL) {
+      return;
     }
+
+    // We did find anything to execute, notify any waiter so they can install a new one.
+    ml_op_lock.notify_all();
+    ml_op_lock.wait(GuaranteedSafepointInterval);
```

Comment:
I initially had the same concern as in my previous review: if there are multiple threads entering here, they could just ping-pong each other with notify-wait. But then I realized this method is only called from VMThread itself. Maybe methods like that need to be `assert(is_VM_thread())` on entry?

Reply:
Adding assert(s)!